library/core/src/iter/adapters/dedup.rs

```diff
@@ -50,9 +50,7 @@ where
     }
 
     fn size_hint(&self) -> (usize, Option<usize>) {
-        let min = self.last.as_ref().map(|_| 1).unwrap_or(0);
-        let max = self.inner.size_hint().1;
-        (min, max)
+        if self.last.is_some() { (1, self.inner.size_hint().1) } else { (0, Some(0)) }
```

Comment:
Well, it doesn't fix my last comment.
You can have `self.last.is_some()` true and `self.inner.size_hint().1` returning `Some(0)` which result in (1, Some(0)).
I think, you should use something like your previous code:
```rust
let from_stored = self.last.as_ref().map(|_| 1).unwrap_or(0);
let inner_upper = self.inner.size_hint().1;
(from_stored, inner_upper.map(move|x|x+from_stored))
```

Reply:
In the [itertools](https://docs.rs/itertools/0.10.0/src/itertools/adaptors/coalesce.rs.html#59) crate, the following is used:
```rust
let (low, hi) = size_hint::add_scalar(self.iter.size_hint(), self.last.is_some() as usize);
((low > 0) as usize, hi)
```
where `add_scalar` is defined like [this](https://github.com/rust-itertools/itertools/blob/master/src/size_hint.rs#L25).